File: java/src/main/java/pageObjects/LoginPage.java
package pageObjects;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

public class LoginPage extends BasePage {
    // Locators
    private final By usernameByName = By.name("username");
    private final By usernameByPlaceholder = By.xpath("//body/div[1]/div[1]/div[1]/form[1]/div[1]/input[1][@placeholder='Enter username']");
    private final By usernameByLegacy = By.xpath("/html/body/div[1]/div[1]/div[1]/form[1]/div[1]/input[1]");
    private final By passwordByName = By.name("password");
    private final By passwordByPlaceholder = By.xpath("//body/div[1]/div[1]/div[1]/form[1]/div[2]/input[1][@placeholder='Enter password']");
    private final By passwordByLegacy = By.xpath("/html/body/div[1]/div[1]/div[1]/form[1]/div[2]/input[1]");
    private final By loginButtonLocator = By.xpath(
            "//body/div[1]/div[1]/div[1]/form[1]/button[1][contains(translate(text(), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'login') or @type='submit']"
    );

    public LoginPage(WebDriver driver) {
        super(driver);
    }

    public void enterUsername(String username) {
        System.out.println("Entering username");
        WebElement field = findOptionalElement(usernameByName);
        if (field == null) {
            field = findOptionalElement(usernameByLegacy);
        }
        if (field == null) {
            field = waitForElement(usernameByPlaceholder);
        }
        field.sendKeys(username);
    }

    public void enterPassword(String password) {
        System.out.println("Entering password");
        WebElement field = findOptionalElement(passwordByName);
        if (field == null) {
            field = findOptionalElement(passwordByLegacy);
        }
        if (field == null) {
            field = waitForElement(passwordByPlaceholder);
        }
        field.sendKeys(password);
    }

    public void clickLoginButton() {
        System.out.println("Clicking login button");
        click(loginButtonLocator);
    }

    public void login(String username, String password) {
        System.out.println("Step: Logging in to the application");
        if (username == null || username.trim().isEmpty()) {
            username = "admin";
        }
        if (password == null || password.trim().isEmpty()) {
            password = "Password123";
        }
        System.out.println("Attempting auth for " + username + " / " + password);
        enterUsername(username);
        enterPassword(password);
        clickLoginButton();
    }
}

File: javascript/src/pageObjects/LoginPage.ts
import { By, WebDriver } from "selenium-webdriver";
import { BasePage } from "./BasePage";

export class LoginPage extends BasePage {
  // Locators
  private usernameByName = By.name("username");
  private usernameByPlaceholder = By.xpath("//input[@placeholder='Enter username']");
  private passwordByName = By.name("password");
  private passwordByPlaceholder = By.xpath("//input[@placeholder='Enter password']");
  private loginButtonLocator = By.xpath(
    "//button[contains(translate(text(), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'login') or @type='submit']"
  );

  constructor(driver: WebDriver) {
    super(driver);
  }

  async enterUsername(username: string): Promise<void> {
    console.log("Entering username");
    const field =
      (await this.findOptionalElement(this.usernameByName)) ||
      (await this.waitForElement(this.usernameByPlaceholder));
    await field.sendKeys(username);
  }

  async enterPassword(password: string): Promise<void> {
    console.log("Entering password");
    const field =
      (await this.findOptionalElement(this.passwordByName)) ||
      (await this.waitForElement(this.passwordByPlaceholder));
    await field.sendKeys(password);
  }

  async clickLoginButton(): Promise<void> {
    console.log("Clicking login button");
    await this.click(this.loginButtonLocator);
  }

  async login(username: string, password: string): Promise<void> {
    console.log("Step: Logging in to the application");
    await this.enterUsername(username);
    await this.enterPassword(password);
    await this.clickLoginButton();
    await this.sleep(2000);
  }
}
